src/SignalR/clients/csharp/Client.Core/src/HubConnection.cs

```diff
@@ -646,36 +646,54 @@ async Task OnStreamCanceled(InvocationRequest irq)
         {
             Dictionary<string, object> readers = null;
             streamIds = null;
-            var newArgs = new List<object>(args.Length);
-
+            int newArgsCount = args.Length;
+            Span<bool> isStreaming = args.Length < 1024
+                ? stackalloc bool[args.Length]
```

Comment:
Perf-hint: it is better to stack-allocate to a constant value* then slice to the appropriate size.

```c#
Span<bool> isStreaming = args.Length <= 1024
    ? stackalloc bool[1024].Slice(args.Length)
    : new bool[args.Length];
```

The value of 1024 seems quite large to me. Is 256 enough (at least to cover the very common cases)?

\* results in less machine-code to figure out the correct size of the stackspace needed as it's already know. If it's a power of two even better.

Reply:
Ah that's a good point! I was following [Dos and Dont's of stackalloc](https://vcsjones.dev/stackalloc/), and didn't quite catch the Slicing benefit.

Would the above be `Slice(0, args.Length)`, or `Slice(args.Length)`?

Regarding the size I took 256 and pushed it out to 1024 to hit 1KB. If it is more beneficial to have a power of 2, and 256 seems like a reasonable boundary I can change that.

For my understanding is the constant value performance gained on a level below .NET? I created a quick sample in [SharpLab.io](https://sharplab.io/#v2:C4LglgNgPgAgTARgLACgYGYAE9MGFMDeqmJmAxgPYB2AzsJmFfQLICGAHgMrCtkDWnMAC8AppgC8mBAAY4AFgDcxUhmxzMzABSN6rAE4BzGgEpCy0qU4AHVlQA8AIwCewEQD5MNG1QmZ9RzA82Lh5+QVFzCyiAfkwqEQB3TGdXAG1/GgBdSKjSEE9QvlYICAoyZJcRVODuXgFhEUylFAsAX0jI1Rh1ADltJj9DEzMW3MxrW0dKjy9bXwzAjQ5asIacsdj4pJSqjOzRsbyCuuLS8p3q5cLwxoA6TggwMhFNaQAaQaNjZraOg/JqHQGAMatcGnBfHAAKwANh+Kiw3UwAHl+rohqYiP8LBN7DsZt55kNFqC6jc4OtcptEhU0ntKVF8nQTiUyrSqqTVqI4Jl7o9nq8PhlvpF2ihWkA==), and I didn't see a difference in the IL that was generated.

Reply:
Can we try an implementation without stackalloc to see how it compares?

Also, it looks like you're relying on stackalloc being zero-initialized which isn't something you should rely on.

Reply:
> Would the above be `Slice(0, args.Length)`, or `Slice(args.Length)`?

`Slice(0, args.Length)` (sorry that the comment shows it wrong).

> For my understanding is the constant value performance gained on a level below IL?

The JIT can produce better [machine code](https://sharplab.io/#v2:EYLgxg9gTgpgtADwGwBYA0AXEBDAzgWwB8ABAJgEYBYAKGIAYACY8gOgCUBXAOwwEt8YLAMIR8AB14AbGFADKMgG68wMXAG4aNYgGYG3XNgBmMJqQYAxCBAYBvGgwdNdwAJ4YTAQQAUvHg1y8AF4wAJT2jnbUjtEMru4AVAxiGFAMALz+GNhgANbYkpIQYLFuMADaAcEAuhpRMQ7EAOwM8clQtdEAvuEOPU4l7gwAQj5+laF9kfUOcTCJbemZ2XkFRQPlpACsSDV90U0tbR2O3XW9Zwxlsjm8YgAyRfm4AJJcvBhVfTrrDEKjGP4gjBAmELlNprN5ilFrgsrl8oVirMylsdsd6gdWil0QxTp0gA=) as with the constant (a power of 2) it's just a pointer move and zeroing -- except `[SkipLocalsInit]` then it's only the pointer move.
If stack-allocated to a `size` then the JIT has emit code that figures out the right size at runtime, which is overhead as it needs to round up to a power of two anyway.
Further note: small constant value (i think 16 is the limit) will be treated as ordinary locals by the JIT -- another optimization done.

> relying on stackalloc being zero-initialized

Good point!
Right now you can rely on the runtime to zero-init the stack-allocated content, as there's no `SkipLocalsInit`-attribute applied. But be aware of https://github.com/dotnet/aspnetcore/issues/26586.

Reply:
All super helpful information, thanks! I updated the [BenchmarkDotNet gist](https://gist.github.com/joncloud/b6b8ae1c4f5fc63c3c7821006d695e62) to include several more iterations:

* With256 - Swaps out to only `stackalloc` a constant size (256) and slice to the target args.
* WithSkipLocalsInit - Based on `With256`, and uses `SkipLocalsInit` along with an `else` block to initialize the span values.
* WithAllocOnly - Swaps out `stackalloc` and always uses `new` to create the `isStreaming` array. @BrennanConroy is this what you were intending?

Reply:
>     * @BrennanConroy is this what you were intending?

No, I meant not allocating a new list until you found a stream item. Would probably make the code a bit complicated.

Your results for Update vs. With256 are odd, in almost all cases 256 is slower which doesn't really make sense.

Also, you can probably just use `.Clear()` instead of the `else` condition for `SkipLocalsInit`.

Reply:
I pushed up another commit using the path for stackalloc + SkipLocalsInit. This included enabling the `<AllowUnsafeBlocks>` flag in `Microsoft.AspNetCore.SignalR.Client.Core.csproj`.

@BrennanConroy a couple of things:
* Do you think it is worthwhile to investigate conditionally allocating the list based on streaming arguments?
* I spun up an Azure B4ms VM (4 vCPUs, 16GiB RAM, Debian 10.7), ran the benchmarks and the results for `Update` vs. `With256` were pretty consistent.
* I also ran [some small tests](https://gist.github.com/joncloud/0f4934def0c1c292eccdb877d66c44ab), which tested my initialization vs. `Span<T>.Clear()` and it appears that in this situation the `else` block may be more efficient. Though that said, I may have something implemented wrong.

Let me know if there is anything I can do.

Reply:
This is excellent work @joncloud!

Reply:
I'm mostly happy with this. Although I'm leaning towards removing the `SkipLocalsInit` so we don't have to enable unsafe code. The gain's look very small for enabling it.

Reply:
Yes remove it.

Reply:
Done deal! Anything else I can do?